### PR TITLE
Preserve query key order in All Products

### LINKED ISSUE #1609: All Products: Changing back to "sort by Latest" shows no results
**Describe the bug**
In the front end, the All Products block can show no results when you change the sorting from `Latest` to something else and back again. (See gif below.)

Discovered when testing #1598

**To Reproduce**
1. Ensure you have some products. Publish a post with All Products block and default settings (sort by newness/latest).
2. View the post on front end. Note there are products displayed.
3. Select a different sort from the sort by dropdown, e.g. `Average rating`.
4. Re-select sort by `Latest`. 

- All products shows `No products` and placeholder image for first page of results.
- Other results pages show products as normal.

**Expected behavior**
Show products on first results page, consistent with before switching sort order.

**Screenshots**

sort-latest-disappear

**WordPress (please complete the following information):**

* Core version: 5.3.2
* Gutenberg version (if installed): Plugin not installed
* WooCommerce version: 3.8.1
* WooCommerce Blocks version: 2.5.10
* Site language: English
* Any other plugins installed: Jetpack 8.1
* Theme: Storefront 2.5.3

**Desktop (please complete the following information):**

* OS: macOS
* Browser: Safari
* Version: 13.0.4 (15608.4.9.1.3)

### PR DESCRIPTION
**Pull request by Aljullu**
Fixes #1609.

#### tl;dr

Running this code snippet in your browser console:
```JS
var products1 = { a: 'WordPress', b: 'Jetpack', c: 'WooCommerce' };
var products2 = { b: 'Tumblr', d: 'Akismet' };
console.log( { ...products1, ...products2 } );
```
Outputs a different result depending on the browser:
* Safari & other Webkit browsers: `{a: "WordPress", c: "WooCommerce", b: "Tumblr", d: "Akismet"}`.
* Firefox & Chrome: `{ a: "WordPress", b: "Tumblr", c: "WooCommerce", d: "Akismet" }`.

That was the cause of #1609: Firefox & Chrome preserve key order when doing a merge while Webkit, doesn't.

In the _All Products_ block, when changing the sort select, we are [updating the query merging two objects](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/base/hooks/use-query-state.js#L125-L128) and later we [stringify the `query` object](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/data/collections/selectors.js#L23) and use that stringified-query to know [if that query is already in the store state](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/data/collections/selectors.js#L24).

So, when the query is updated (for example, when changing the order), the attributes change their order in Webkit, making the default query never to be matched again and, because of that, returning 0 results.

One solution would have been to migrate the query data structure to a [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), which preserves key order. But that required a big refactor that I don't think we want in a point release. So for now I relied on using lodash's `assign` method that allows merging objects while preserving key order.

### How to test the changes in this Pull Request:

1. Create a post with an _All Products_ block.
2. In a Webkit browser (Safari, for example), change the sort select to something different and then go back to the initial order.
3. Verify results appear, instead of the _No results_ message.

### Changelog

> Fix bug in Safari and other Webkit browsers that was causing the All Products block to show 0 results when resetting the sort value.

### CODE REVIEW

**Comment by senadir** [comment]
Great find! thank you for working on this Albert! I just don’t understand this bit

>So, when the query is updated (for example, when changing the order), the attributes change their order in Webkit, making the default query never to be matched again and, because of that, returning 0 results.

Shouldn’t this cause a new fetch of data?

**Comment by Aljullu** [comment]
> Shouldn’t this cause a new fetch of data?

@senadir good question! Somebody with more experience in `@wordpress/data` might be able to explain it better than me, but from my understanding the issue is that we call the selector with the `query` object, while internally it's saved in the state with the stringified `query` as the key. Shallow-comparing the `query` object returned true, so it thought it had already fetched the resources, but when getting the values from the state, it was comparing the stringified `query`, which was different.

So it was not fetching the resource again neither finding it in the current store state. That's why it was showing 0 results.

**Comment by nerrad** [comment]
In a roundabout way, Albert's explanation is correct :)
